src/L2/resolver/InterfaceResolver.sol

```diff
@@ -0,0 +1,90 @@
+
+import {ResolverBase} from "./ResolverBase.sol";
+
+/// @title Interface Resolver
+///
+/// @notice ENSIP-8 compliant Interface Discovery resolver. Adaptation of the ENS InterfaceResolver.sol profile contract, with
+///         EIP-7201 storage compliance.
+///         https://github.com/ensdomains/ens-contracts/blob/staging/contracts/resolvers/profiles/InterfaceResolver.sol
+///
+/// @author Coinbase (https://github.com/base-org/basenames)
+abstract contract InterfaceResolver is IInterfaceResolver, ResolverBase {
+    struct InterfaceResolverStorage {
+        /// @notice Interface implementer address by interface id, node and version.
+        mapping(uint64 version => mapping(bytes32 node => mapping(bytes4 interfaceId => address implemenentor)))
+            versionable_interfaces;
+    }
+
+    /// @notice EIP-7201 storage location.
+    // keccak256(abi.encode(uint256(keccak256("interface.resolver.storage")) - 1)) & ~bytes32(uint256(0xff));
+    bytes32 constant INTERFACE_RESOLVER_STORAGE = 0x933ab330cd660334bb219a68b3bfaf86387ecd49e4e53a39e8310a5bd6910c00;
+
+    /// @notice Sets an interface implementer address.
+    ///
+    /// @dev Setting the address to 0 restores the default behaviour of querying the contract at `addr()` for interface support.
+    ///
+    /// @param node The node to update.
+    /// @param interfaceID The EIP-165 interface ID.
+    /// @param implementer The address of a contract that implements this interface for this node.
+    function setInterface(bytes32 node, bytes4 interfaceID, address implementer) external virtual authorised(node) {
+        _getInterfaceResolverStorage().versionable_interfaces[_getResolverBaseStorage().recordVersions[node]][node][interfaceID]
+        = implementer;
+        emit InterfaceChanged(node, interfaceID, implementer);
+    }
+
+    /// @notice Returns the address of a contract that implements the specified interface for this name.
+    ///
+    /// @dev If an implementer has not been set for this interfaceID and name, this resolver will query
+    ///     itself at `addr(node)`. If `addr()` is set, a contract exists at that address, and that
+    ///     contract implements EIP165 and returns `true` for the specified interfaceID, its address
+    ///     will be returned.
+    /// @param node The ENS node to query.
+    /// @param interfaceID The EIP 165 interface ID to check for.
+    ///
+    /// @return The address that implements this interface, or address(0) if the interface is unsupported.
+    function interfaceImplementer(bytes32 node, bytes4 interfaceID) external view virtual override returns (address) {
+        address implementer = _getInterfaceResolverStorage().versionable_interfaces[_getResolverBaseStorage()
+            .recordVersions[node]][node][interfaceID];
+        if (implementer != address(0)) {
+            return implementer;
+        }
+
+        address a = IAddrResolver(address(this)).addr(node);
```

Comment:
NOTE: This is a deviation from ENS's inheritance-based approach and should be looked at closely.